src/addrman.h

```diff
@@ -229,13 +229,48 @@ friend class CAddrManTest;
     //! Holds addrs inserted into tried table that collide with existing entries. Test-before-evict discipline used to resolve these collisions.
     std::set<int> m_tried_collisions;
 
+    void CheckWithLockHeld() EXCLUSIVE_LOCKS_REQUIRED(cs)
+    {
+#ifdef DEBUG_ADDRMAN
+        AssertLockHeld(cs);
+        if (const int err = Check_(); err != 0) {
+            LogPrintf("ADDRMAN CONSISTENCY CHECK FAILED!!! err=%i\n", err);
+        }
+#endif // DEBUG_ADDRMAN
+    }
+
+    void ClearWithLockHeld() EXCLUSIVE_LOCKS_REQUIRED(cs)
```

Comment:
It'd be good to get rid of the `Clear()` functions entirely:

- the call to `Clear()` in deserialize could be replaced with an assert that size() is 0 (we should never deserialize into a populated `CAddrman`
- the calls in the tests can be replaced by instantiating new `CAddrman`s. Clearing then reusing an existing `CAddrman` is not a realistic usage pattern.

If you don't want to do that in this PR, it feels like this function should be called `Clear_()` and be defined in the cpp file, for consistency with the other functions.

Reply:
@jnewbery 
> It'd be good to get rid of the `Clear()` functions entirely:

I totally agree with you.

>  - the call to `Clear()` in deserialize could be replaced with an assert that size() is 0 (we should never deserialize into a populated `CAddrman`)

Going to implement this suggestion.

>  - the calls in the tests can be replaced by instantiating new `CAddrman`s. Clearing then reusing an existing `CAddrman` is not a realistic usage pattern.

This part is not trivial, and probably deserves its own pr.

Reply:
Thanks! [Updated](https://github.com/bitcoin/bitcoin/pull/19238#issuecomment-846605462).